### DISCUSSION TITLE
Retrieve multiple secrets in that have same keys

### DISCUSSION THREAD

**Discussion by jobas2007** [awspring/spring-cloud-aws#388] [Q&A]
label:"component: secrets-manager"

** also did put on stack overflow - https://stackoverflow.com/questions/72264774/retrieve-multiple-aws-secrete-in-spring-boot-app-that-have-same-keys
Have a spring boot app (with starter parent at 2.4.8).

The app is connecting to multiple data sources, and the automated Jenkins job being used to create secrets across company does it such a way that although secrets names are different(per app/source) but they all have same value underneath - 'username' and 'password' text.

For example:

First secret would be '/secret/rds/dev/foo/foo_app_user' with value:

username : user1
password: pass1

second secret would be '/secret/snowflake/dev/hoo/hoo_app_user' with value:

username : user2
password: pass2

Trying to figure out how can they both be imported using spring config import, while still being able to be used distinctively in properties/yaml file (NOTE: already have tried with optional also)

`
```
spring:
    config:
         import: aws-secretsmanager:/secret/rds/dev/foo/foo_app_user,aws-secretsmanager:/secret/snowflake/dev/hoo/hoo_app_user
```

...
```
system:
   cache:
       username: ${username}
       password: ${password}

```
....
```
snowflake:
  datasource: 
     username: ${username}
     password: ${password}
```

**Comment by maciejwalkowiak** [comment]
This isn't currently possible and I am not sure if ever will be. Do other `spring.config.import` implementations support it?

**Reply by mdreem** [reply]
In my project we are facing the same issue and I think it is essential that this is possible somehow. Maybe I am understanding the Secretsmanager wrong, but it seems quite suggestive to have different secrets for e.g. different databases and then I still would like to just call the entries still "username" and "password". Currently they would be overridden and the secret name itself also lends itself to appear in the property.
Similar to the ParameterStore where this actually happens. It surely feels a bit different as there is the recursive search call.

As I need it anyway and in the worst case will probably implement an internal library to handle it, I tried to prototype something along the lines of the ParameterStore: https://github.com/mdreem/spring-cloud-aws/commit/e987eb1bc25c70983c6b7d3edb58e1929eea867d

For this I added a parameter named "spring.cloud.aws.secretsmanager.prefix" that triggers the namespacing and also some stripping along the lines of the ParameterStore behavior.

It is not optimal yet, but maybe helpful to start a discussion, to see if there is a viable solution to this.

**Comment by abdelghanymarzouk** [comment]
Hi @jobas2007,

I'm facing the same problem, did you find a solution?

**Comment by maciejwalkowiak** [comment]
To summarize - the issue is that when multiple secrets containing JSON properties are imported, and same keys exist in both secret JSONs then value from one secret overwrites the other.

For example

- `/secret/first-secret`:

```json
{ "username": "postgres" }
```

- `/secret/second-secret`:
```json
{ "username": "root" }
```

I am thinking how this can be solved. We could add a prefix option:

```yml
spring:
    config:
         import: aws-secretsmanager:/secrets/first-secret?prefix=first,aws-secretsmanager:/secret/second-secret?prefix=second
```

Then they would be imported as:

```
first.username=postgres
second.username=root
```

What do you think about it? Perhaps there is more idiomatic way to do it with Spring Boot? @spencergibb @wilkinsona your suggestion would be appreciated.

**Reply by abdelghanymarzouk** [reply]
I think this is a very good idea.

**Reply by wilkinsona** [reply]
There's nothing built in to do this but there are some building blocks that you could use. The requirement seems somewhat similar to the [support for configuring a system environment variable prefix](https://github.com/spring-projects/spring-boot/wiki/Spring-Boot-2.5-Release-Notes#environment-variable-prefixes) that we [added in Boot 2.5](https://github.com/spring-projects/spring-boot/issues/3450). You may be able to use `org.springframework.boot.context.properties.source.ConfigurationPropertySource.withPrefix(String)` that was introduced as part of those changes in your `aws-secretsmanager` implementation.

**Reply by maciejwalkowiak** [reply]
Thanks @wilkinsona!

**Comment by maciejwalkowiak** [comment]
We have PR ready https://github.com/awspring/spring-cloud-aws/pull/622

**Reply by abdelghanymarzouk** [reply]
Thanks for your changes, kindly have you already released it? if yes which version is it?

**Reply by maciejwalkowiak** [reply]
It will be released in 3.0 RC1 hopefully next week

**Reply by snigdhasjg** [reply]
Can we make the same change for `2.4.x`? Looking at the code, I don't see any challenge. Also verified in my local.

**Reply by snigdhasjg** [reply]
upgrading to cloud `3.0.x` brings lots of breaking changes because we will have to upgrade spring `3.0.x`

**Reply by maciejwalkowiak** [reply]
We are not actively working on 2.x anymore but if you contribute a PR with a change including updating reference docs we are happy to merge.

**Reply by snigdhasjg** [reply]
Added the change #630

Please verify the docs, as this is 1st time I'm submitting PR on this repo. Not sure abt convention followed

**Reply by abdelghanymarzouk** [reply]
Hello @snigdhasjg @maciejwalkowiak,

Kindly is there any idea when it will be merged?

**Reply by maciejwalkowiak** [reply]
When @MatejNedic is back ;-) Jokes aside, we plan for making releases next week.

**Reply by abdelghanymarzouk** [reply]
Good to know :D  thanks

**Reply by snigdhasjg** [reply]
Thanks folks, it got merged 😇

**Reply by abdelghanymarzouk** [reply]
thanks all, has it been released? or not yet?

**Reply by maciejwalkowiak** [reply]
Not yet. I recommend watching repository for releases it should happen in upcoming days.
